docs/graphql.md

```diff
@@ -0,0 +1,153 @@
+
+We considered hashing queries to tell them apart, but decided against it.
+Instead we will just consider all unnamed queries _unnamed_ and consequently put them in the same bucket.
+
+<details>
+<summary>Rationale:</summary>
+
+1. Some clients generate `id`s from hashing the contents of the query (see [apollo-tooling](https://github.com/apollographql/apollo-tooling/blob/1dfd737eaf85b89b2cfb13913342e091e3c03d18/packages/apollo-codegen-core/src/compiler/visitors/generateOperationId.ts#L5)). This would split the anonymous queries into separate buckets.
+
+    A problem with this approach is that a user of the APM UI has no way to recognise queries in the transactions list before clicking through.
+
+    Using just the `id` will not reveal the true culprit since there can be variables associated with the query. Different values for the variables can lead to very different workloads and response times.
+
+2. Another approach is to simply label them `[unnamed]`.
+
+    A problem with _that_ approach is that the contents and thereby the relevant db queries and other sub-span actions that the server might do while resolving these queries may be wildly different making it hard to provide a _true_ sample waterfall.
+
+    These two examples for example will look the same for the top-level GraphQL spans but will represent significantly different workloads.
+
+    ```
+    [- anonymous graphql span --------------]
+      [- 1,000x SELECT * ---------------]
+        [- 1,000 more SELECT * -]
+
+    [- anonymous graphql span --------------]
+      [- SELECT id FROM users WHERE id=? -]
+    ```
+
+No one of these are perfect. Because the benefits of using `id`s in the worst case could be misleading anyway, we're going with option 2.
+</details>
+
+To further help and nudge developers to use Operation Names for their queries, a tooltip will be shown in the Kibana UI, suggesting to use Operation Names.
+
+Transaction name examples:
+- `GraphQL: [unnamed]`
+
+### Batching/Multiplexing queries
+
+Some clients and servers allow batching/multiplexing queries (see for example [apollo-link-batch-http](https://www.apollographql.com/docs/link/links/batch-http/#gatsby-focus-wrapper) or [dataloader](https://github.com/graphql/dataloader#batching)) allowing multiple queries to be run from the same HTTP request.
+
+If multiple queries are run from the same request, we join their operation names in the transaction name with a `+`.
+
+Transaction name examples:
+- `GraphQL: UserWithComments+PostWithSiblings+MoreThings+[unnamed]`
+
+To avoid very long transaction names, if a request has more than five queries, we abbreviate it to `[more-than-five-queries]`.
+
+Transaction name examples:
+- `GraphQL: [more-than-five-queries]`
+
+## Spans
+
+The life cycle of responding to a GraphQL query is mostly split in two parts. First, the quer(y/ies) are read, parsed and analyzed. Second, they are executed.
+
+![Example GraphQL waterfall](graphql-example.png)
```

Comment:
forgot to checkin the image?

Reply:
It’s here even though the path might need to be absolute url to actually work. 


![](https://github.com/elastic/apm/blob/dd45b26ed41acc1071aa0096002979095c4a41f3/docs/graphql-example.png)